### DISCUSSION TITLE
No redirection when the session expires (Authelia & Caddy)

### DISCUSSION THREAD

**Discussion by EctoBoggan** [authelia/authelia#6758] [Q&A] [answered]
Hello,

I'm using Authelia with Caddy to redirect a service (ytdl.mydomain.com)

After 5 minutes, my cookie session expires correctly but i'm not redirected to authelia auth panel when trying to access my service again (by pression F5 or opening it in a new tab), it just allow me to access a non-working only visual aspect of my service. I have to SHIFT + F5 / CTRL + F5 to refresh and clear my tab cache for it to redirect me to Authelia auth panel. (I tried with chrome, firefox and edge)

This is how I configured my session :

```
access_control:
  default_policy: deny
  rules:
    - domain: ytdl.mydomain.com
      policy: two_factor

session:
  name: authelia_session  
  same_site: lax
  secret: xxxxxxxxxxxxxxx
  expiration: 5m
  inactivity: 5m
  domain: mydomain.com
  remember_me_duration:  1y
```

This how i'm redirecting my service using caddy :

```
ytdl.mydomain.com {
    forward_auth 192.168.0.200:9091 {
        uri /api/verify?rd=https://auth.mydomain.com/
        copy_headers Remote-User Remote-Groups Remote-Name Remote-Email
    }
    reverse_proxy 192.168.0.200:6021 {
        header_down -Access-Control-Allow-Origin
    }
}
```

Caddy isn't showing any error or any log concerning that but Authelia does : 

```2024/02/29 14:14:08	stderr	time=\2024-02-29T14:14:08+01:00\" level=info msg=\"Access to https://ytdl.mydomain.com/custom-service-worker.js (method GET) is not authorized to user <anonymous>	 responding with status code 302 with location redirect to https://auth.mydomain.com/?rd=https%3A%2F%2Fytdl.mydomain.com%2Fcustom-service-worker.js&rm=GET\" method=GET path=/api/verify remote_ip=172.17.0.1
			
2024/02/29 14:14:08	stderr	time=\2024-02-29T14:14:08+01:00\" level=info msg=\"Access to https://ytdl.mydomain.com/ngsw-worker.js (method GET) is not authorized to user <anonymous>	 responding with status code 302 with location redirect to https://auth.mydomain.com/?rd=https%3A%2F%2Fytdl.mydomain.com%2Fngsw-worker.js&rm=GET\" method=GET path=/api/verify remote_ip=172.17.0.1
			
2024/02/29 14:14:08	stderr	time="2024-02-29T14:14:08+01:00" level=debug msg="Check authorization of subject username= groups= ip=172.17.0.1 and object https://ytdl.mydomain.com/ngsw-worker.js (method GET)."	
			
2024/02/29 14:14:08	stderr	time="2024-02-29T14:14:08+01:00" level=debug msg="Check authorization of subject username= groups= ip=172.17.0.1 and object https://ytdl.mydomain.com/custom-service-worker.js (method GET)."	
			
2024/02/29 14:14:07	stderr	time=\2024-02-29T14:14:07+01:00\" level=info msg=\"Access to https://ytdl.mydomain.com/socket.io/?EIO=4&transport=polling&t=Otr03DE (method GET) is not authorized to user <anonymous>	 responding with status code 302 with location redirect to https://auth.mydomain.com/?rd=https%3A%2F%2Fytdl.mydomain.com%2Fsocket.io%2F%3FEIO%3D4%26transport%3Dpolling%26t%3DOtr03DE&rm=GET\" method=GET path=/api/verify remote_ip=172.17.0.1
			
2024/02/29 14:14:07	stderr	time="2024-02-29T14:14:07+01:00" level=debug msg="Check authorization of subject username= groups= ip=172.17.0.1 and object https://ytdl.mydomain.com/socket.io/?EIO=4&transport=polling&t=Otr03DE (method GET)."	
			
2024/02/29 14:14:03	stderr	time=\2024-02-29T14:14:03+01:00\" level=info msg=\"Access to https://ytdl.mydomain.com/socket.io/?EIO=4&transport=polling&t=Otr02An (method GET) is not authorized to user <anonymous>	 responding with status code 302 with location redirect to https://auth.mydomain.com/?rd=https%3A%2F%2Fytdl.mydomain.com%2Fsocket.io%2F%3FEIO%3D4%26transport%3Dpolling%26t%3DOtr02An&rm=GET\" method=GET path=/api/verify remote_ip=172.17.0.1
			
2024/02/29 14:14:03	stderr	time="2024-02-29T14:14:03+01:00" level=debug msg="Check authorization of subject username= groups= ip=172.17.0.1 and object https://ytdl.mydomain.com/socket.io/?EIO=4&transport=polling&t=Otr02An (method GET)."	
```

I'll be happy to give more info if needed.
Have a nice day

**Comment by glime-ai** [comment]
Please note this response was created using generative AI leveraging previously available information from the repository. This information may be helpful however it is our strong recommendation that you carefully consider the information before blindly applying it.

---

Based on the information provided and the configuration snippets for both Authelia and Caddy, it seems like the issue might be related to how the browser handles the session expiration and caching, particularly for service workers and socket.io connections. When the session expires, Authelia correctly identifies that the user is no longer authorized and attempts to redirect to the login page. However, due to the browser's caching behavior, especially with service workers (`custom-service-worker.js` and `ngsw-worker.js`) and dynamic content fetched through `socket.io`, the browser might not immediately recognize that it needs to refresh the page from the server to get the new redirect to the login page.

Service workers, in particular, can aggressively cache content to enable offline functionality and fast loading times, which might interfere with the expected redirect behavior when the session expires. Similarly, socket.io connections might maintain a persistent connection that doesn't immediately reflect the change in authentication status.

To address this issue, you might consider the following approaches:

1. **Service Worker Update and Activation**: Ensure that your service worker is configured to update itself and activate the new version immediately upon changes. This might involve listening to the `install` event in the service worker and calling `self.skipWaiting()`, and then controlling page refreshes or cache updates on the `activate` event. This approach can help ensure that the service worker doesn't serve cached content that might interfere with the redirect after session expiration.

2. **Cache Management for Authenticated Content**: Review and adjust the caching strategies for content that requires authentication, particularly for dynamic content fetched through `socket.io`. You might need to implement cache-busting techniques or adjust the cache headers to prevent the browser from serving stale content after the session expires.

3. **Explicitly Handling Session Expiration**: On the client side, you could implement explicit checks for session expiration, especially for applications that rely heavily on JavaScript and dynamic content fetching. This might involve intercepting fetch requests or socket.io messages to check for authentication failures and then forcing a page reload or redirect to the login page when necessary.

4. **Debugging and Testing**: Use browser developer tools to inspect the service worker behavior, cache storage, and network requests. This can help identify exactly how the cached content is being served and how the service worker is interacting with your application after session expiration.

Unfortunately, without direct access to the service worker code and the specific implementation details of your application, these suggestions are somewhat general. Adjusting the behavior of service workers and managing browser cache effectively can be complex and might require specific adjustments based on your application's architecture and requirements.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=6fsHkcyufjQLaB&utm_medium=github).

**Comment by james-d-elliott** [comment]
Sounds like you're checking the remember me box on login in which case it's working as intended.

**Reply by EctoBoggan** [reply]
Thank you for your reply,
Unfortunately, I don't check it (I tried and it works, the cookie expiration date is correctly set to one year).
But with both options (checked and unchecked), when the cookie expires, the page is still reachable but is "frozen", I'm not redirected to Authelia authentication automatically, I have to refresh the page by clearing the cache (SHIFT + F5 or CTRL + F5 on FireFox).

Could this be caused by not using Redis and/or using users_database.yml instead of a real database?

I'll be happy to provide more informations,
Have a nice day

**Reply by james-d-elliott** [reply]
Based on this additional information this is most likely a PWA which is not making new HTTP requests either due to cache. Can you replicate it with all apps or just one?

> Could this be caused by not using Redis and/or using users_database.yml instead of a real database?

No they are not related. Redis is used to store session information.

**Reply by EctoBoggan** [reply]
Thank you for your reply,
I tried with another app ([pawelmalak/flame](https://github.com/pawelmalak/flame)) ... and it works correctly. 
When the cookie expires, Authelia asks me to authenticate myself again.

So it's probably a problem with the application itself ([alexta69/metube](https://github.com/alexta69/metube)).

I'll dig further, maybe it would be as as simple as adding a ressource for the app.

Have a nice day.

**Reply by james-d-elliott** [reply]
It might be, the app may be blocking redirects or it may just not be making any additional requests.

**Reply by EctoBoggan** [reply, marked as answer]
I made some changes in my Caddy configuration in order to control how cache behave

```
ytdl.domain.com {
    header {
        Cache-Control "no-cache, no-store, must-revalidate"
    }
    forward_auth 192.168.0.200:9091 {
        uri /api/verify?rd=https://auth.domain.com/
        copy_headers Remote-User Remote-Groups Remote-Name Remote-Email
    }
    reverse_proxy 192.168.0.200:6021
}
```
What I did in order to test : 

1. Set the session expiration to one minute
2. Login with Authelia to access my app
3. Wait one minute
4. Reload the page using F5 or the reload button on my browser

OR

1. Login with Authelia to access my app
2. Delete my cookie
3. Reload the page using F5 or the reload button on my browser

It **works** on all Chromium-based browsers (Chrome, Edge, Brave...) but not on Firefox. (instead of not working on any of these browsers)

So I think it's more a problem related to my proxy header configuration rather than an Authelia configuration.

I'll stop bothering you here since it seems to not concerne Authelia

Thanks a lot for your time !

**Reply by james-d-elliott** [reply]
No stress, if you ever have a question that may not be related it's fine generally for us until people get accusatory or otherwise disorderly. Sometimes we'll just know an answer since we've got a lot of experience or other users may know.
